R/bundlePackagePackrat.R

```diff
@@ -0,0 +1,200 @@
+snapshotPackratDependencies <- function(bundleDir,
+                                        implicit_dependencies = character(),
+                                        verbose = FALSE) {
+
+  addPackratSnapshot(bundleDir, implicit_dependencies, verbose = verbose)
+
+  lockFilePath <- packratLockFile(bundleDir)
+  df <- as.data.frame(read.dcf(lockFilePath), stringsAsFactors = FALSE)
+  unlink(dirname(lockFilePath), recursive = TRUE)
+
+  # get repos defined in the lockfile
+  repos <- gsub("[\r\n]", " ", df[1, "Repos"])
+  repos <- strsplit(unlist(strsplit(repos, "\\s*,\\s*", perl = TRUE)), "=", fixed = TRUE)
+  repos <- setNames(
+    sapply(repos, "[[", 2),
+    sapply(repos, "[[", 1)
+  )
+
+  # get packages records defined in the lockfile
+  records <- utils::tail(df, -1)
+  if (nrow(records) == 0) {
+    return(data.frame())
+  }
+
+  rownames(records) <- NULL
+  records <- records[manifestPackageColumns(records)]
+  records[c("Source", "Repository")] <- standardizeRecords(records, repos)
+
+  records$description <- lapply(records$Package, function(pkg) {
+    readLines(system.file("DESCRIPTION", package = pkg))
+  })
+
+  records
+}
+
+standardizeRecords <- function(records, repos) {
+  availablePackages <- availablePackages(repos)
+  repos <- standardizeRepos(repos)
+
+  rows <- lapply(seq_len(nrow(records)), function(i) {
+    standardizePackratPackage(records[i, ], availablePackages, repos = repos)
+  })
+  rows <- lapply(rows, as.data.frame, stringsAsFactors = FALSE)
+  rbind_fill(rows, c("Source", "Repository"))
+}
+
+standardizeRepos <- function(repos) {
+  # Ensure that each repository has a unique name
+  names(repos) <- ifelse(
+    names2(repos) == "",
+    paste0("repo_", seq_along(repos)),
+    names2(repos)
+  )
+
+  # And strip trailing /
+  repos <- gsub("/$", "", repos)
+
+  repos
+}
+
+standardizePackratPackage <- function(record, availablePackages, repos = character()) {
+  pkg <- record$Package
+  source <- record$Source
+
+  # source types are defined by packrat:
+  # https://github.com/rstudio/packrat/blob/v0.9.0/R/pkg.R#L328
+  if (source %in% c("github", "gitlab", "bitbucket")) {
+    # SCM information is recorded elsewhere
+    repository <- NA_character_
+  } else if (source == "source") {
+    # can't install source packages elsewhere
+    repository <- NA_character_
+    source <- NA_character_
+  } else if (source == "CustomCRANLikeRepository") {
+    # Package was installed from source, but packrat guessed it was installed
+    # from a known repo. We need to check that it's not actually a dev version
+    if (isDevVersion(record, availablePackages)) {
+      repository <- NA_character_
+      source <- NA_character_
+    } else {
+      repository <- findRepoUrl(pkg, availablePackages)
+      source <- findRepoName(repository, repos)
+    }
+  } else if (source %in% c("CRAN", "Bioconductor")) {
```

Comment:
```suggestion
  } else if (source == "CustomCRANLikeRepository" && isDevVersion(record, availablePackages)) {
    # Package was installed from source, but packrat guessed it was installed
    # from a known repo.
      repository <- NA_character_
      source <- NA_character_
  } else if (source %in% c("CRAN", "Bioconductor")) {
```

To me, this makes the overall conditional flow of this block easier to follow -- the nested `else` is the same as the final `else` in the outer set of conditionals. Feel free to take or leave, this is definitely a nitpick and a matter of taste.

Reply:
I like it, thanks!